### PR TITLE
feat: enabling more tests and adding more far-call tests

### PR DESCRIPTION
**Pull request by mm-zk**
# What ❔

Fixing far_call (and other) unittests.

## Why ❔

* To cover the most recent circuit mismatches that we fixed.

### CODE REVIEW

File: src/tests/run_manually.rs [hunk 1]
```diff
@@ -173,18 +174,40 @@ pub(crate) fn run_and_try_create_witness_inner(asm: &str, cycle_limit: usize) {
     run_and_try_create_witness_for_extended_state(bytecode, vec![], cycle_limit)
 }
 
+#[derive(Default, Clone)]
+pub struct Options {
+    // How many cycles shoudl the main VM run for.
```
Comment (StanislavBreadless) [inline src/tests/run_manually.rs]:
```suggestion
    // How many cycles should the main VM run for.
```

File: src/tests/run_manually.rs [hunk 2]
```diff
@@ -173,18 +174,40 @@ pub(crate) fn run_and_try_create_witness_inner(asm: &str, cycle_limit: usize) {
...
+    // If not set - default is 50.
```
Comment (StanislavBreadless) [inline src/tests/run_manually.rs]: nit: maybe move this fifty into a constant? otherwise the comment will get outdated. ofc it is a test, but the change is small also

File: src/tests/simple_tests/asm_tests.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+use zkevm_assembly::Assembly;
+
+use self::run_manually::{run_and_try_create_witness_for_extended_state, Options};
+
+use super::*;
+
+use crate::tests::run_manually::run_and_try_create_witness_inner;
+
+/// Runs the tests based on the ASM files from a given directory.
+/// The main assembly shoudl be in `entry.asm` file, while additional
```
Comment (StanislavBreadless) [inline src/tests/simple_tests/asm_tests.rs]:
```suggestion
/// The main assembly should be in `entry.asm` file, while additional
```

File: src/tests/simple_tests/asm_tests.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            ));
+            let bytecode = Assembly::try_from(asm.to_owned())
+                .unwrap()
+                .compile_to_bytecode()
+                .expect(&format!("Failed to compile {:?}", file_path));
+            (Address::from_low_u64_be(*address as u64), bytecode)
+        })
+        .collect();
+
+    //run_and_try_create_witness_for_extended_state(entry_bytecode, contracts, 500);
```
Comment (StanislavBreadless) [inline src/tests/simple_tests/asm_tests.rs]: commented out code

File: src/tests/simple_tests/testdata/meta_opcode/entry.asm [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    .rodata.cst32
+    .p2align	5
+    .text
+    .globl	__entry
+__entry:
+.main:
+        add 32, r0, r2
+        add 64, r0, r3
+        st.1.inc r2, r3, r2
+        st.2.inc r2, r3, r2
```
Comment (StanislavBreadless) [inline src/tests/simple_tests/testdata/meta_opcode/entry.asm]: better to add a comment why do we need memory access here

File: src/tests/simple_tests/testdata/log_precompile/entry.asm [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        .text
+        .globl	__entry
+    __entry:
+    .main:
+        add 10000, r0, r1
+        add 1000, r0, r10
+        ; put 2 in r2 (sha)
+        add 2, r0, r2
+        ; precompile call - address 0x1, AuxData (additional costs - 0 for now)
+        log.precompile r2, r3, r4
```
Comment (StanislavBreadless) [inline src/tests/simple_tests/testdata/log_precompile/entry.asm:12]: I am not sure it has intended behavior

`log.precompile` actives `log::precompilecall` opcode, which accepts:

- src0 encodes precompile call abi: https://github.com/matter-labs/era-zkevm_opcode_defs/blob/109d9f734804a8b9dc0531c0b576e2a0f55a40de/src/definitions/abi/precompile_call.rs#L6
- src1 encodes precompile aux data (basically cost for executing it): https://github.com/matter-labs/era-zkevm_opcode_defs/blob/109d9f734804a8b9dc0531c0b576e2a0f55a40de/src/definitions/abi/precompile_call.rs#L52

the type of precompile (sha/keccak/etc) is derived from the `this` address when the precompile is executed. 


In other words here:
- src0 r2 has a rather strange value, but since the entry point does not have a valid precompile address it is ignored
- src1 is 0, i.e. no ergs cost is beared

File: src/tests/simple_tests/testdata/log_precompile/entry.asm [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+        near_call r1, @inner, @handler
+        ; We should never get here - as the near_call should panic due to out of gas.
+        ret.panic r0
+    inner:
+        to_l1 r0, r1
+        ; put 2 in r2 (sha)
+        add 2, r0, r2
```
Comment (StanislavBreadless) [inline src/tests/simple_tests/testdata/log_precompile/entry.asm]: similar to the above, the comment is misleading

File: src/tests/simple_tests/testdata/log_precompile/entry.asm [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+        ; Add extra 16k cost (more than this near call has)
+        add 1, r0, r3
+        shl.s 14, r3, r3
+        shl.s 18, r3, r3
```
Comment (StanislavBreadless) [inline src/tests/simple_tests/testdata/log_precompile/entry.asm]: would one shl 32 suffice?

File: src/tests/simple_tests/testdata/log_precompile/entry.asm [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        add 1, r3, r3
+        shl.s 14, r3, r3
+
+        ; precompile call - address 0x1, AuxData (additional costs - 16k for now)
+        log.precompile r2, r3, r4
```
Comment (StanislavBreadless) [inline src/tests/simple_tests/testdata/log_precompile/entry.asm:27]: it seems like aux abi is 2^14 * (2^32 + 1) => 2^14 pubdata bytes and 0 ergs. Is it the intended behavior or I do not read it correctly?

File: src/tests/simple_tests/far_calls.rs [hunk 10]
```diff
@@ -92,18 +108,26 @@ fn test_far_call_and_return_large_data() {
...
         st.2 r2, r3
+        ; create ABI for far_call
+        ; use 2 for forwarding mode (Aux heap)
         add 2, r0, r1
-        shl.s 136, r1, r1
+        shl.s 32, r1, r1
+        ; give 10k gas
+        add 100000, r1, r1
+        shl.s 96, r1, r1
+        ;shl.s 128, r1, r1
```
Comment (StanislavBreadless) [inline src/tests/simple_tests/far_calls.rs]: probably remove commented out code

File: src/tests/run_manually.rs (hunk 2 above)
Comment (mm-zk) [reply src/tests/run_manually.rs]: good idea - done.

File: src/tests/simple_tests/testdata/meta_opcode/entry.asm (hunk 5 above)
Comment (mm-zk) [reply src/tests/simple_tests/testdata/meta_opcode/entry.asm]: I think it is not needed. removed.

File: src/tests/simple_tests/testdata/log_precompile/entry.asm (hunk 6 above)
Comment (mm-zk) [reply src/tests/simple_tests/testdata/log_precompile/entry.asm:12]: thanks, I didn't notice that this is an ABI there.
So how do you pass the information about the type of precompile that you want to call?

File: src/tests/simple_tests/testdata/log_precompile/entry.asm (hunk 8 above)
Comment (mm-zk) [reply src/tests/simple_tests/testdata/log_precompile/entry.asm]: yes.

File: src/tests/simple_tests/testdata/log_precompile/entry.asm (hunk 9 above)
Comment (mm-zk) [reply src/tests/simple_tests/testdata/log_precompile/entry.asm:27]: ((1<<32) + 1 )<<14.
So both 'u32' fields should have 2^14.

File: src/tests/simple_tests/testdata/log_precompile/entry.asm (hunk 7 above)
Comment (mm-zk) [reply src/tests/simple_tests/testdata/log_precompile/entry.asm]: ack.

File: src/tests/simple_tests/testdata/log_precompile/entry.asm (hunk 6 above)
Comment (mm-zk) [reply src/tests/simple_tests/testdata/log_precompile/entry.asm:12]: ok, go it. fixed.

File: src/tests/simple_tests/testdata/log_precompile/entry.asm (hunk 9 above)
Comment (StanislavBreadless) [reply src/tests/simple_tests/testdata/log_precompile/entry.asm:27]: agree, yes

**Review by StanislavBreadless — APPROVED** [review]
